File: Back_SpringBoot/TappedOut/src/main/java/com/jgl/TappedOut/service/ResultService.java
package com.jgl.TappedOut.service;

import java.util.List;

import com.jgl.TappedOut.dto.ResultCreateDTO;
import com.jgl.TappedOut.dto.ResultResponseDTO;
import com.jgl.TappedOut.dto.ResultUpdateDTO;

/**
 * Interface to declare methods needed at {@link ResultServiceImpl}
 * 
 * @author Jorge García López
 * @version 1.0
 * @since 2025
 */
public interface ResultService {
    List<ResultResponseDTO> getAllResults();
    List<ResultResponseDTO> getResultsByEvent(Long eventId);
    List<ResultResponseDTO> getResultsByEventAndCategory(Long eventId, Long categoryId);
    List<ResultResponseDTO> getResultsByEventAndCompetitor(Long eventId, Long competitorId);
    List<ResultResponseDTO> getResultsByCompetitor(Long competitorId);
    List<ResultResponseDTO> getResultsByEventAndPosition(Long eventId, Integer position);
    ResultResponseDTO getResultById(Long id);
    ResultResponseDTO getWinnerByEventAndCategory(Long eventId, Long categoryId);
    ResultResponseDTO createResult(ResultCreateDTO dto);
    ResultResponseDTO updateResult(Long id, ResultUpdateDTO dto);
    void deleteResult(Long id);
}

File: Front_React/tappedout-front/src/service/resultService.js
/* eslint-disable no-useless-catch */
import apiClient from "../api/apiClient";
import { toast } from "react-hot-toast";

const resultService = {
    getAllResults: async () => {
        try {
            const response = await apiClient.get('/result');
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    getResultsByEvent: async (eventId) => {
        try {
            const response = await apiClient.get(`/result/event/${eventId}`);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    getResultsByEventAndCategory: async (eventId, categoryId) => {
        try {
            const response = await apiClient.get(`/result/event/${eventId}/category/${categoryId}`);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    getResultsByCompetitorAndEvent: async (competitorId, eventId) => {
        try {
            const response = await apiClient.get(`/result/competitor/${competitorId}/event/${eventId}`);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    getResultsByCompetitor: async (competitorId) => {
        try {
            const response = await apiClient.get(`/result/competitor/${competitorId}`);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    getResultsByEventAndPosition: async (eventId, position) => {
        try {
            const response = await apiClient.get(`/result/event/${eventId}/position/${position}`);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    getResultById: async (id) => {
        try {
            const response = await apiClient.get(`/result/${id}`);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    getWinnerByEventAndCategory: async (eventId, categoryId) => {
        try {
            const response = await apiClient.get(`/result/event/${eventId}/category/${categoryId}/winners`);
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    createResult: async (resultData) => {
        try {
            const response = await apiClient.post('/result', resultData);
            toast.success('Result created successfully.');
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    updateResult: async (id, resultData) => {
        try {
            const response = await apiClient.put(`/result/${id}`, resultData);
            toast.success('Result updated successfully.');
            return response.data;
        } catch (error) {
            throw error;
        }
    },

    deleteResult: async (id) => {
        try {
            const response = await apiClient.delete(`/result/${id}`);
            toast.success('Result deleted successfully.');
            return response.data;
        } catch (error) {
            throw error;
        }
    },
};

export default resultService;